src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceParameterService.cs

```diff
@@ -0,0 +1,153 @@
+                var invocationSemanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var editor = new SyntaxEditor(await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false), generator);
+
+                foreach (var invocationExpression in invocationExpressionList)
+                {
+                    var newArgumentExpression = expression;
+                    var invocationArguments = invocationExpression.ArgumentList.Arguments;
+                    foreach (var argument in invocationArguments)
+                    {
+                        var associatedParameter = semanticFacts.FindParameterForArgument(invocationSemanticModel, argument, cancellationToken);
+                        var parenthesizedArgumentExpression = generator.AddParentheses(argument.Expression, false);
+                        if (!mappingDictionary.TryGetValue(associatedParameter, out var value))
+                        {
+                            continue;
+                        }
+
+                        newArgumentExpression = newArgumentExpression.ReplaceNode(newArgumentExpression.GetCurrentNode(value), parenthesizedArgumentExpression);
+                    }
+                    var allArguments = invocationExpression.ArgumentList.Arguments.Add(SyntaxFactory.Argument(newArgumentExpression.WithoutAnnotations(ExpressionAnnotationKind).WithAdditionalAnnotations(Simplifier.Annotation)));
+                    editor.ReplaceNode(invocationExpression, editor.Generator.InvocationExpression(invocationExpression.Expression, allArguments));
+                }
+
+                var newRoot = editor.GetChangedRoot();
+                document = document.WithSyntaxRoot(newRoot);
+                currentSolution = document.Project.Solution;
+            }
+
+            return currentSolution;
+        }
+
+        /// <summary>
+        /// Ties the identifiers within the expression back to their associated parameter
+        /// </summary>
+        /// <param name="expression"> The expression containing identifiers we are tying back to the parameters </param>
+        /// <param name="methodSymbol"> The method containing the expression </param>
+        /// <returns></returns>
+        private static Dictionary<IParameterSymbol, IdentifierNameSyntax> TieExpressionToParameters(ExpressionSyntax expression, IMethodSymbol methodSymbol)
+        {
+            var nameToParameterDict = new Dictionary<IParameterSymbol, IdentifierNameSyntax>();
+            var variablesInExpression = expression.DescendantNodes().Where(node => node is IdentifierNameSyntax);
+
+            foreach (var variable in variablesInExpression)
+            {
+                foreach (var parameter in methodSymbol.Parameters)
+                {
+                    if ((string)((IdentifierNameSyntax)variable).Identifier.Value == parameter.Name)
```

Comment:
1. the cast seems unnecessary.
2. this won't work (if you have locals with teh same names as parameters).

You need ot use the semantic model to see what the identifiers actually bind to.

Reply:
What's a good way to get the symbols in an expression? The one that comes to mind is DataFlow analysis, but I feel there has to be an easier way for just referenced symbols.

Reply:
How do you have a local with the same name as a parameter? Captures?

Reply:
Changed, now done semantically